.github/workflows/proof.yml

```diff
@@ -21,7 +21,7 @@ on:
       - aarch64
 
 jobs:
-  all:
+  proofs:
     name: All
```

Comment:
This is the action for pull request checks. Do we want to trigger binary verification from those? I'd have thought those workflows are too expensive for that.

Reply:
I don't really mind either way.

On one hand, I was thinking that decompilation is possibly cheap enough to run in the same cases we run Isabelle proofs, so we could just try it and see. Decompilation duration is about the same order of magnitude as the AWS proofs. Although decompilation will do the work every time, whereas Isabelle is smart enough to use cached images when possible. On the back end, we can use the tags to control which jobs actually run graph-refine. Initially, I would allow tags `l4v/weekly-clean` and `l4v/proof-deploy/*`, so we wouldn't run graph-refine on PRs unless I intervene to add the job to the queue.

On the other hand, we can manually trigger decompilation on the artifacts produced by PR proof runs, as long as the proof run exports the build artifacts. So we wouldn't lose much by not triggering decompilation here.

What do you think?

Reply:
If it's on the same order as the AWS proofs that means it'll roughly double the PR wait time, because the decompilation can only start after the proof workflow is finished and has uploaded artifacts. I'd says that's too long to always trigger.

But a manual trigger for BV on PRs would make sense, so that we can kick it off if we're doing anything we think might interfere with decompilation (most stuff on Refine and above won't really have any influence on BV outcomes).

We should in fact probably add such a trigger to seL4 pull requests, because they are much more likely to break BV. But let's do that once the infrastructure works here.

Reply:
Ok, so I vote for keeping the artifact upload, but I would remove the automatic trigger for this workflow.

Reply:
Decompilation will be a separate workflow (on the graph-refine repo), triggered from an l4v proof run using repository-dispatch. So it won't delay l4v proof results.

However, having decompilation triggered automatically might mean you can run graph-refine sooner, if you decide you want that. Currently, it would not be possible to manually trigger a decompilation workflow until the l4v proof run is finished, so there would always be a delay between the l4v workflow finishing and the human remembering to start decompilation. We could get around that by making the repository-dispatch conditional on a label on the l4v workflow, but I currently don't have an implementation for that.

So I think the choice is mainly about how much of GitHub's compute budget we want to spend on decompilation, since they've made it free for us (for now!).

Note that strictly speaking, it would be possible to run decompilation in parallel with l4v proofs, but doing that makes some things difficult: It's harder to ensure that l4v and the decompiler are using the same kernel build. And it's harder to figure out how to trigger graph-refine, since it depends on both decompilation and l4v. The choice to have decompilation triggered from l4v is largely to make those problems go away.

W.r.t. triggers on seL4 pull requests, my thought was that we would use the existing mechanisms to trigger l4v proof runs, and then the l4v proof workflow would trigger decompilation. We could use a label on the seL4 PR to signal whether the l4v proof run should trigger decompilation, but the overall process would still need to go via the l4v proof workflow.

Reply:
> Decompilation will be a separate workflow (on the graph-refine repo), triggered from an l4v proof run using repository-dispatch. So it won't delay l4v proof results.

It wouldn't delay those results, but if we wanted to know if it succeeded or not, we'd still have to wait. If we don't want to know, we probably shouldn't run it. While the minutes are free, the number of concurrent runners is limited, which we'd be using up.

> Currently, it would not be possible to manually trigger a decompilation workflow until the l4v proof run is finished, so there would always be a delay between the l4v workflow finishing and the human remembering to start decompilation. We could get around that by making the repository-dispatch conditional on a label on the l4v workflow, but I currently don't have an implementation for that.

Yes, I was thinking of the label approach. It's Ok to leave that for now, it's more important to get coverage for a release. And I think I can implement the label relatively easily from my side if need be.

> W.r.t. triggers on seL4 pull requests, my thought was that we would use the existing mechanisms to trigger l4v proof runs, and then the l4v proof workflow would trigger decompilation. We could use a label on the seL4 PR to signal whether the l4v proof run should trigger decompilation, but the overall process would still need to go via the l4v proof workflow.

This works for pushes to master (after merge), but not for pull request. Proofs on pull requests in the seL4 repo are currently just `CRefine` and `SimplExportAndRefine` and run in the seL4 repo. But if the infrastructure is available to run it in l4v with repository dispatch, I can do the rest when we decide we want that feature.

Reply:
> While the minutes are free, the number of concurrent runners is limited, which we'd be using up.

That's a good point, which I hadn't considered. So I agree that we should remove it here.

> > Currently, it would not be possible to manually trigger a decompilation workflow until the l4v proof run is finished, so there would always be a delay between the l4v workflow finishing and the human remembering to start decompilation. We could get around that by making the repository-dispatch conditional on a label on the l4v workflow, but I currently don't have an implementation for that.
> 
> Yes, I was thinking of the label approach. It's Ok to leave that for now, it's more important to get coverage for a release. And I think I can implement the label relatively easily from my side if need be.

Sounds good.

> > W.r.t. triggers on seL4 pull requests, my thought was that we would use the existing mechanisms to trigger l4v proof runs, and then the l4v proof workflow would trigger decompilation. We could use a label on the seL4 PR to signal whether the l4v proof run should trigger decompilation, but the overall process would still need to go via the l4v proof workflow.
> 
> This works for pushes to master (after merge), but not for pull request. Proofs on pull requests in the seL4 repo are currently just `CRefine` and `SimplExportAndRefine` and run in the seL4 repo. But if the infrastructure is available to run it in l4v with repository dispatch, I can do the rest when we decide we want that feature.

Right, that makes sense, and now I see what you meant by your original comment. I'll make a PR to add the artifact upload to the seL4 proof workflow, but I'll leave out the trigger for now. The manual decompilation trigger will be able to retrieve kernel build artifacts from the seL4 proof workflow in the same way it gets them from l4v proof workflows.